cylc/flow/simulation.py

```diff
@@ -16,25 +16,116 @@
 """Utilities supporting simulation and skip modes
 """
 
+from dataclasses import dataclass
 from typing import TYPE_CHECKING, Any, Dict, List, Optional, Union
 from time import time
 
 from cylc.flow.cycling.loader import get_point
-from cylc.flow.network.resolvers import TaskMsg
+from cylc.flow.parsec.util import (
+    pdeepcopy,
+    poverride
+)
 from cylc.flow.platforms import FORBIDDEN_WITH_PLATFORM
 from cylc.flow.task_state import (
     TASK_STATUS_RUNNING,
     TASK_STATUS_FAILED,
     TASK_STATUS_SUCCEEDED,
 )
-from cylc.flow.wallclock import get_current_time_string
+from cylc.flow.wallclock import get_unix_time_from_time_string
 
 from metomi.isodatetime.parsers import DurationParser
 
 if TYPE_CHECKING:
-    from queue import Queue
-    from cylc.flow.cycling import PointBase
+    from cylc.flow.broadcast_mgr import BroadcastMgr
+    from cylc.flow.task_events_mgr import TaskEventsManager
     from cylc.flow.task_proxy import TaskProxy
+    from cylc.flow.workflow_db_mgr import WorkflowDatabaseManager
+    from cylc.flow.cycling import PointBase
+
+
+@dataclass
+class ModeSettings:
+    """A store of state for simulation modes.
+
+    Used instead of modifying the runtime config.
+
+    Args:
+        itask:
+            The task proxy this submission relates to.
+        broadcast_mgr:
+            The broadcast manager is used to apply any runtime alterations
+            pre simulated submission.
+        db_mgr:
+            The database manager must be provided for simulated jobs
+            that are being resumed after workflow restart. It is used to
+            extract the original scheduled finish time for the job.
+
+    Attrs:
+        simulated_run_length:
+            The length of time this simulated job will take to run in seconds.
+        timeout:
+            The wall-clock time at which this simulated job will finish as
+            a Unix epoch time.
+        sim_task_fails:
+            True, if this job is intended to fail when it finishes, else False.
+
+    """
+    simulated_run_length: float = 0.0
+    sim_task_fails: bool = False
+    timeout: float = 0.0
+
+    def __init__(
+        self,
+        itask: 'TaskProxy',
+        broadcast_mgr: 'BroadcastMgr',
+        db_mgr: 'WorkflowDatabaseManager',
+    ):
+
+        # itask.summary['started_time'] and mode_settings.timeout need
+        # repopulating from the DB on workflow restart:
+        started_time = itask.summary['started_time']
+        try_num = None
+        if started_time is None:
+            # Get DB info
+            db_info = db_mgr.pub_dao.select_task_job(
```

Comment:
Ooh, is this supposed to be the public database actually?

Reply:
I'm sorry, I don't understand what you are asking.

Reply:
> Do you think I _should_ be consulting the private database?

Yes I would have thought so

Reply:
Yes, this should use the private DB, good spot.

Reply:
```suggestion
            db_info = db_mgr.pri_dao.select_task_job(
```